src/halmos/__main__.py

```diff
@@ -963,38 +966,38 @@ def is_unknown(result: CheckSatResult, model: Model) -> bool:
     return result == unknown or (result == sat and not is_model_valid(model))
 
 
+def refine(query: str) -> str:
+    # replace uninterpreted abstraction with actual symbols for assertion solving
+    # TODO: replace `(evm_bvudiv x y)` with `(ite (= y (_ bv0 256)) (_ bv0 256) (bvudiv x y))`
+    #       as bvudiv is undefined when y = 0; also similarly for evm_bvurem
+    query = re.sub(r"(\(\s*)evm_(bv[a-z]+)(_[0-9]+)?\b", r"\1\2", query)
+    return re.sub(
+        r"\(\s*declare-fun\s+evm_(bv[a-z]+)(_[0-9]+)?\b",
+        r"(declare-fun dummy_\1\2",
+        query,
+    )
+
+
 def gen_model(args: Namespace, idx: int, ex: Exec) -> ModelWithContext:
     if args.verbose >= 1:
         print(f"Checking path condition (path id: {idx+1})")
 
-    model = None
-
     ex.solver.set(timeout=args.solver_timeout_assertion)
     res = ex.solver.check()
-    if res == sat:
-        model = ex.solver.model()
+    model = ex.solver.model() if res == sat else None
 
-    if is_unknown(res, model) and args.solver_fresh:
+    if res == sat and not is_model_valid(model):
         if args.verbose >= 1:
-            print(f"  Checking again with a fresh solver")
-        sol2 = SolverFor("QF_AUFBV", ctx=Context())
-        # sol2.set(timeout=args.solver_timeout_assertion)
-        sol2.from_string(ex.solver.to_smt2())
-        res = sol2.check()
-        if res == sat:
-            model = sol2.model()
-
-    if is_unknown(res, model) and args.solver_subprocess:
+            print(f"  Checking again with refinement")
+        res, model = solve(refine(ex.solver.to_smt2()), args)
+
+    if args.solver_subprocess and is_unknown(res, model):
```

Comment:
does `--solver-subprocess` still serve a purpose here? it may return a sat/unsat even if we got unknown previously because it runs without a timeout?

Reply:
yes, and you can run a different solver using this option too.